### PR TITLE
Graphic Buffers Update

### PR DESCRIPTION
**Pull request by bremco**
The highly interdependent Issues this fixes:
- The Vertex Array Object (VAO) for a render wasn't actually being (re)used, but regenerated each time. Now it's cached.
- While (otherwise) vertex- and index-buffers where already cached for normal rendering, the index-buffer for a mesh was recreated each time only a _part_ of the mesh (a range) needed to be drawn (for example, Layer and Simulation view in Cura). (NB: This was due to a workaround on our side, caused by an only _partial_ implementation of a specific bit of the OpenGL API by the Python interface we're using.) This is now fixed.

Complications:
- In order to show 'ranged'/partial meshes, implementing programs (like Cura) must now have shaders that take a draw-range uniform, which can be compared against the newly introduced 'vertex-index' attribute. (As we can't use the OpenGL API due to partial implementation, and we can't use the `gl_primitiveID` built-in due to support of legacy shaders.) See also https://github.com/Ultimaker/Cura/pull/10188, **the front-end implementation of this PR.**

See also this related discussion https://github.com/Ultimaker/Uranium/issues/664 earlier.

### CODE REVIEW

**Comment by nallath** [comment]
Do you also have before / after profiling? I'm curious what the gains are :)

**Comment by bremco** [comment]
@nallath Right, was one of the things I meant to do as well :-)
Not sure I can get that done today though.

(P.S. Even if there are no true gains I'd argue this is more correct, but that's harder to justify from a 'risk' perspective. Also technically upping the SDK with my shader shenanigans.)

File: UM/View/RenderBatch.py [hunk 1]
```diff
@@ -234,43 +229,7 @@ def render(self, camera: Optional[Camera]):
...
-            index_buffer = OpenGL.getInstance().createIndexBuffer(mesh)
-        else:
-            # glDrawRangeElements does not work as expected and did not get the indices field working..
-            # Now we're just uploading a clipped part of the array and the start index always becomes 0.
-            index_buffer = OpenGL.getInstance().createIndexBuffer(
-                mesh, force_recreate=True, index_start = self._render_range[0], index_stop = self._render_range[1])
-        if index_buffer is not None:
-            index_buffer.bind()
-
+    def _setMeshAttributes(self, mesh: Any):
```
Comment (nallath) [inline UM/View/RenderBatch.py:232]:
```suggestion
    def _setMeshAttributes(self, mesh: Any) -> None:
```
Partially typed functions are not checked by type checking.

File: UM/View/RenderBatch.py [hunk 2]
```diff
@@ -302,21 +261,77 @@ def _renderItem(self, item: Dict[str, Any]):
...
-                else:
-                    self._gl.glDrawElements(self._render_mode, mesh.getFaceCount(), self._gl.GL_UNSIGNED_INT, None)
-            else:
-                if self._render_mode == self.RenderMode.Triangles:
-                    self._gl.glDrawRangeElements(self._render_mode, self._render_range[0], self._render_range[1], self._render_range[1] - self._render_range[0], self._gl.GL_UNSIGNED_INT, None)
-                else:
-                    self._gl.glDrawElements(self._render_mode, self._render_range[1] - self._render_range[0], self._gl.GL_UNSIGNED_INT, None)
-        else:
-            self._gl.glDrawArrays(self._render_mode, 0, vertex_count)
+    def _vertexBuffersSetup(self, mesh: Any):
```
Comment (nallath) [inline UM/View/RenderBatch.py:264]:
```suggestion
    def _vertexBuffersSetup(self, mesh: Any) -> None:
```

File: UM/View/RenderBatch.py [hunk 3]
```diff
@@ -302,21 +261,77 @@ def _renderItem(self, item: Dict[str, Any]):
...
+            index_buffer.bind()
+
+        self._setMeshAttributes(mesh)
+
+        # Cache and return:
+        mesh.setCachedUserValue(self._shader.getReferenceKey(), vao)
+        vao.release()
+        return vao
+
+    def _renderItem(self, item: Dict[str, Any]):
```
Comment (nallath) [inline UM/View/RenderBatch.py:296]:
```suggestion
    def _renderItem(self, item: Dict[str, Any]) -> None:
```

**Comment by bremco** [comment]
Lesson for next time: Run the tests _after_ I've merged with main as well.

**Comment by bremco** [comment]
(... actually I don't think I did run the tests. I was confused with last week, when I had to set them up again on my work lappy.)

In any case, please try again, should pass now!

File: UM/View/RenderBatch.py [hunk 4]
```diff
@@ -302,21 +261,79 @@ def _renderItem(self, item: Dict[str, Any]):
...
-                else:
-                    self._gl.glDrawElements(self._render_mode, mesh.getFaceCount(), self._gl.GL_UNSIGNED_INT, None)
-            else:
-                if self._render_mode == self.RenderMode.Triangles:
-                    self._gl.glDrawRangeElements(self._render_mode, self._render_range[0], self._render_range[1], self._render_range[1] - self._render_range[0], self._gl.GL_UNSIGNED_INT, None)
-                else:
-                    self._gl.glDrawElements(self._render_mode, self._render_range[1] - self._render_range[0], self._gl.GL_UNSIGNED_INT, None)
-        else:
-            self._gl.glDrawArrays(self._render_mode, 0, vertex_count)
+    def _vertexBuffersSetup(self, mesh: Any) -> Optional[QOpenGLVertexArrayObject]:
```
Comment (fieldOfView) [inline UM/View/RenderBatch.py:264]: `Any` is a bit lazy here, when you obviously expect `MeshData`

Comment (bremco) [reply UM/View/RenderBatch.py]: Oops! Fixed in https://github.com/Ultimaker/Uranium/pull/714/commits/8611188f88ec11dbef37bcd40dd96f79d17540cd

There's a joke about lazy evaluation in there ... _but I just couldn't be bothered._

**Comment by bremco** [comment]
I still didn't do a proper benchmark, but playing around a bit with some printouts of the average time of the render of the SimulationView (where I expect most of the time-win will be) it seems that the newer code is about 20% faster.

File: UM/View/RenderBatch.py [hunk 5]
```diff
@@ -234,43 +229,7 @@ def render(self, camera: Optional[Camera]):
...
-            index_buffer = OpenGL.getInstance().createIndexBuffer(mesh)
-        else:
-            # glDrawRangeElements does not work as expected and did not get the indices field working..
-            # Now we're just uploading a clipped part of the array and the start index always becomes 0.
-            index_buffer = OpenGL.getInstance().createIndexBuffer(
-                mesh, force_recreate=True, index_start = self._render_range[0], index_stop = self._render_range[1])
-        if index_buffer is not None:
-            index_buffer.bind()
-
+    def _setMeshAttributes(self, mesh: Any) -> None:
```
Comment (fieldOfView) [inline UM/View/RenderBatch.py:232]:
```suggestion
    def _setMeshAttributes(self, mesh: MeshData) -> None:
```

**Comment by bremco** [comment]
When trying to benchmark this, I noticed `lineMeshVertexCount` (the entire layer version, not the polygon one), which I now use in SimulationPass, is quite a slow function compared to what was there before.

I need to set up some caching or something, or this PR will feel incomplete (even though it's already a bit faster). A lot of SimilationPass could be imporved, but let's try to keep it a bit smaller for now.

**Comment by rburema** [comment]
Closing this in favour of https://github.com/Ultimaker/Uranium/pull/721 due to laptop & account related shenanigans.

**Pull request closed without merging**